types/coin.go

```diff
@@ -226,6 +246,40 @@ func (coins Coins) safePlus(coinsB Coins) Coins {
 	}
 }
 
+// SubCoinByDenom looks for a matching coin in coins that has the same denom as
+// other. Upon matching, the other coin is subtracted from the matching coin. If
+// the resulting coin is zero, it is removed. If the coins are empty, the coin
+// is simply added to the empty set.
+func (coins Coins) SubCoinByDenom(other Coin) Coins {
+	res := copyCoins(coins)
+
+	if len(res) == 0 {
+		return Coins{other}
+	}
+
+	for i, coin := range res {
+		if coin.Denom == other.Denom {
+			otherNeg := Coin{other.Denom, other.Amount.Neg()}
+
+			res[i] = coin.Plus(otherNeg)
+			if res[i].IsZero() {
+				// remove resulting zero coin
+				res = append(res[:i], res[i+1:]...)
+			} else if !res[i].IsNotNegative() {
```

Comment:
this is weird because it will panic for "0" correct? we should probably replace with an `IsNegative()` function I imagine

Reply:
Hmmm, if `res[i]` is 0 then it'll be removed (line #265) or am I missing/misunderstanding something?

Reply:
I think you're right but I also find the double negation and conditional-overlap confusing to read, why not just use `IsNegative()` ?

Reply:
Ahh, that is because we do not have a `IsNegative`. I shall add this.

Reply:
Updated.